Local install: settings.json hooks use relative paths, break after any `cd`

### GSD Version

1.34.2

### Runtime

Claude Code

### Operating System

macOS

### Node.js Version

v25.9.0

### Shell

/bin/zsh

### Installation Method

npx get-shit-done-cc@latest (fresh run)

### What happened?

> Disclaimer: This bug was detected by a human programmer, but the report itself is AI-generated. I've reviewed it thoroughly, though my JavaScript experience is limited. I don't want to waste open-source maintainers' time by asking you to review an AI-written report — so please let me know if more effort is needed from my side for you to look into this.

After every Bash tool call whose persistent shell `cwd` is not the project root (e.g. left there by a previous `cd subdir && …`), Claude Code displays one or more of:

```
PreToolUse:Bash hook error
PostToolUse:Bash hook error
```

The tools still execute (the hooks are non-blocking), but the noise is constant — Claude Code persists the bash tool's `cwd` between calls, so a single `cd` early in a session permanently breaks every Pre/PostToolUse hook for the rest of that session.

The same applies to PreToolUse:Edit / PostToolUse:Write etc. — every hook entry written by the local installer is affected, not just `Bash`.

### What did you expect?

Hooks should execute regardless of which subdirectory the user (or a previous tool call) is currently in. Claude Code provides `$CLAUDE_PROJECT_DIR` for exactly this case, and the official Claude Code hooks guide recommends anchoring hook commands to it.

### Steps to reproduce

1. `cd /path/to/any-project`
2. Install GSD locally (any method — npm, source clone running `node bin/install.js`, or `npx get-shit-done-cc@latest`); choose Claude Code, **local** install
3. Inspect `.claude/settings.json` — every hook command should look like:
   ```json
   "command": "node .claude/hooks/gsd-context-monitor.js"
   ```
   (a bare relative path, no `$CLAUDE_PROJECT_DIR` prefix)
4. Open the project in Claude Code
5. Run any Bash tool that changes the persistent shell `cwd`, e.g.:
   ```bash
   cd subdir && uv run python whatever.py
   ```
6. Run any subsequent Bash tool, e.g. `pwd`
7. Observe `PreToolUse:Bash hook error` and `PostToolUse:Bash hook error` appended to the tool output
8. Run the failing hook commands manually from `subdir/`:
   ```bash
   bash .claude/hooks/gsd-validate-commit.sh
   # bash: .claude/hooks/gsd-validate-commit.sh: No such file or directory
   # exit 127

   node .claude/hooks/gsd-context-monitor.js
   # Error: Cannot find module '/.../subdir/.claude/hooks/gsd-context-monitor.js'
   # MODULE_NOT_FOUND, exit 1
   ```

### Error output / logs

```shell
# Reproduced live in a project where `subdir/` is a subdirectory of the project root:


$ pwd
/path/to/project/subdir

$ bash .claude/hooks/gsd-validate-commit.sh
bash: .claude/hooks/gsd-validate-commit.sh: No such file or directory
exit=127

$ node .claude/hooks/gsd-context-monitor.js
node:internal/modules/cjs/loader:1478
  throw err;
  ^
Error: Cannot find module '/path/to/project/subdir/.claude/hooks/gsd-context-monitor.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1475:15)
    ...
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
exit=1

# The hooks themselves are fine — they execute cleanly when invoked from the project root. The failure is purely path resolution.
```

### GSD Configuration

```json

```

### GSD State (if relevant)

```markdown

```

### Runtime settings.json (if relevant)

```json

```

### How often does this happen?

Every time (100% reproducible)

### Impact

Major — Core feature is broken, no workaround

### Workaround (if any)

Manually rewrite `.claude/settings.json` to anchor every hook command to `$CLAUDE_PROJECT_DIR`. For each entry, replace e.g.:

```json
"command": "node .claude/hooks/gsd-context-monitor.js"
```

with:

```json
"command": "node \"$CLAUDE_PROJECT_DIR\"/.claude/hooks/gsd-context-monitor.js"
```

(and the equivalent for `bash` entries). Affects all 6 hook commands the local installer writes:

- `gsd-check-update.js` (SessionStart)
- `gsd-session-state.sh` (SessionStart)
- `gsd-context-monitor.js` (PostToolUse)
- `gsd-phase-boundary.sh` (PostToolUse)
- `gsd-prompt-guard.js` (PreToolUse)
- `gsd-read-guard.js` (PreToolUse)
- `gsd-workflow-guard.js` (PreToolUse)
- `gsd-validate-commit.sh` (PreToolUse)
- `gsd-statusline.js` (statusLine)

The patch is reverted on the next `npx get-shit-done-cc` / installer run.

### Additional context

## Root cause

`bin/install.js` builds hook command strings differently for global vs local installs. For **global** installs it calls `buildHookCommand(targetDir, …)` which produces an absolute path. For **local** installs it concatenates `dirName` with the script name and writes the bare relative string straight into `settings.json` — no `$CLAUDE_PROJECT_DIR`, no `cwd:` field, no anchoring. From [`bin/install.js` line 5547](https://github.com/gsd-build/get-shit-done/blob/v1.34.2/bin/install.js#L5547) on the v1.34.2 tag:

```js
const contextMonitorCommand = isGlobal
  ? buildHookCommand(targetDir, 'gsd-context-monitor.js')
  : 'node ' + dirName + '/hooks/gsd-context-monitor.js';
```

The same pattern repeats for all 9 hook entries (statusline, update-check, context-monitor, prompt-guard, read-guard, workflow-guard, session-state, phase-boundary, validate-commit). Claude Code persists the bash tool's `cwd` across tool calls, so once any tool does `cd subdir && …`, every subsequent hook resolves against `<subdir>/.claude/hooks/...` and fails.

## Suggested fix

Prefix every local-install hook command with `"$CLAUDE_PROJECT_DIR"/`:

```js
const contextMonitorCommand = isGlobal
  ? buildHookCommand(targetDir, 'gsd-context-monitor.js')
  : 'node "$CLAUDE_PROJECT_DIR"/' + dirName + '/hooks/gsd-context-monitor.js';
```

`$CLAUDE_PROJECT_DIR` is set by Claude Code precisely to make hook commands `cwd`-independent.

### Privacy Checklist

- [x] I have reviewed all pasted output for PII (usernames, paths, API keys) and redacted where necessary

--- a/bin/install.js
+++ b/bin/install.js
@@ -5538,21 +5538,24 @@ function install(isGlobal, runtime = 'claude') {
     return;
   }
   const settings = validateHookFields(cleanupOrphanedHooks(rawSettings));
+  // Local installs anchor paths to $CLAUDE_PROJECT_DIR so hooks resolve
+  // correctly regardless of the shell's current working directory (#1906).
+  const localPrefix = '"$CLAUDE_PROJECT_DIR"/' + dirName;
   const statuslineCommand = isGlobal
     ? buildHookCommand(targetDir, 'gsd-statusline.js')
-    : 'node ' + dirName + '/hooks/gsd-statusline.js';
+    : 'node ' + localPrefix + '/hooks/gsd-statusline.js';
   const updateCheckCommand = isGlobal
     ? buildHookCommand(targetDir, 'gsd-check-update.js')
-    : 'node ' + dirName + '/hooks/gsd-check-update.js';
+    : 'node ' + localPrefix + '/hooks/gsd-check-update.js';
   const contextMonitorCommand = isGlobal
     ? buildHookCommand(targetDir, 'gsd-context-monitor.js')
-    : 'node ' + dirName + '/hooks/gsd-context-monitor.js';
+    : 'node ' + localPrefix + '/hooks/gsd-context-monitor.js';
   const promptGuardCommand = isGlobal
     ? buildHookCommand(targetDir, 'gsd-prompt-guard.js')
-    : 'node ' + dirName + '/hooks/gsd-prompt-guard.js';
+    : 'node ' + localPrefix + '/hooks/gsd-prompt-guard.js';
   const readGuardCommand = isGlobal
     ? buildHookCommand(targetDir, 'gsd-read-guard.js')
-    : 'node ' + dirName + '/hooks/gsd-read-guard.js';
+    : 'node ' + localPrefix + '/hooks/gsd-read-guard.js';
 
   // Enable experimental agents for Gemini CLI (required for custom sub-agents)
   if (isGemini) {
@@ -5705,7 +5708,7 @@ function install(isGlobal, runtime = 'claude') {
     // /gsd-quick or /gsd-fast for state-tracked changes. Advisory only.
     const workflowGuardCommand = isGlobal
       ? buildHookCommand(targetDir, 'gsd-workflow-guard.js')
-      : 'node ' + dirName + '/hooks/gsd-workflow-guard.js';
+      : 'node ' + localPrefix + '/hooks/gsd-workflow-guard.js';
     const hasWorkflowGuardHook = settings.hooks[preToolEvent].some(entry =>
       entry.hooks && entry.hooks.some(h => h.command && h.command.includes('gsd-workflow-guard'))
     );
@@ -5730,7 +5733,7 @@ function install(isGlobal, runtime = 'claude') {
     // Configure commit validation hook (Conventional Commits enforcement, opt-in)
     const validateCommitCommand = isGlobal
       ? 'bash ' + targetDir.replace(/\\/g, '/') + '/hooks/gsd-validate-commit.sh'
-      : 'bash ' + dirName + '/hooks/gsd-validate-commit.sh';
+      : 'bash ' + localPrefix + '/hooks/gsd-validate-commit.sh';
     const hasValidateCommitHook = settings.hooks[preToolEvent].some(entry =>
       entry.hooks && entry.hooks.some(h => h.command && h.command.includes('gsd-validate-commit'))
     );
@@ -5757,7 +5760,7 @@ function install(isGlobal, runtime = 'claude') {
     // Configure session state orientation hook (opt-in)
     const sessionStateCommand = isGlobal
       ? 'bash ' + targetDir.replace(/\\/g, '/') + '/hooks/gsd-session-state.sh'
-      : 'bash ' + dirName + '/hooks/gsd-session-state.sh';
+      : 'bash ' + localPrefix + '/hooks/gsd-session-state.sh';
     const hasSessionStateHook = settings.hooks.SessionStart.some(entry =>
       entry.hooks && entry.hooks.some(h => h.command && h.command.includes('gsd-session-state'))
     );
@@ -5779,7 +5782,7 @@ function install(isGlobal, runtime = 'claude') {
     // Configure phase boundary detection hook (opt-in)
     const phaseBoundaryCommand = isGlobal
       ? 'bash ' + targetDir.replace(/\\/g, '/') + '/hooks/gsd-phase-boundary.sh'
-      : 'bash ' + dirName + '/hooks/gsd-phase-boundary.sh';
+      : 'bash ' + localPrefix + '/hooks/gsd-phase-boundary.sh';
     const hasPhaseBoundaryHook = settings.hooks[postToolEvent].some(entry =>
       entry.hooks && entry.hooks.some(h => h.command && h.command.includes('gsd-phase-boundary'))
     );

--- a/tests/bug-1906-hook-relative-paths.test.cjs
+++ b/tests/bug-1906-hook-relative-paths.test.cjs
@@ -1,0 +1,82 @@
+/**
+ * Regression tests for bug #1906
+ *
+ * Local installs must anchor hook command paths to $CLAUDE_PROJECT_DIR so
+ * hooks resolve correctly regardless of the shell's current working directory.
+ *
+ * The original bug: local install hook commands used bare relative paths like
+ * `node .claude/hooks/gsd-context-monitor.js`. Claude Code persists the bash
+ * tool's cwd between calls, so a single `cd subdir && …` early in a session
+ * permanently broke every hook for the rest of that session.
+ *
+ * The fix prefixes all local hook commands with "$CLAUDE_PROJECT_DIR"/ so
+ * path resolution is always anchored to the project root.
+ */
+
+'use strict';
+
+const { describe, test, before } = require('node:test');
+const assert = require('node:assert/strict');
+const fs = require('fs');
+const path = require('path');
+
+const INSTALL_SRC = path.join(__dirname, '..', 'bin', 'install.js');
+
+// All hooks that the installer registers for local installs
+const HOOKS = [
+  'gsd-statusline.js',
+  'gsd-check-update.js',
+  'gsd-context-monitor.js',
+  'gsd-prompt-guard.js',
+  'gsd-read-guard.js',
+  'gsd-workflow-guard.js',
+  'gsd-validate-commit.sh',
+  'gsd-session-state.sh',
+  'gsd-phase-boundary.sh',
+];
+
+describe('bug #1906: local hook commands use $CLAUDE_PROJECT_DIR', () => {
+  let src;
+
+  before(() => {
+    src = fs.readFileSync(INSTALL_SRC, 'utf-8');
+  });
+
+  test('localPrefix variable is defined with $CLAUDE_PROJECT_DIR', () => {
+    assert.match(src, /const localPrefix\s*=\s*['"]\"\$CLAUDE_PROJECT_DIR['"]\s*\//,
+      'localPrefix should be defined using $CLAUDE_PROJECT_DIR');
+  });
+
+  for (const hook of HOOKS) {
+    test(`${hook} local command uses localPrefix (not bare dirName)`, () => {
+      // Find all local command strings for this hook
+      // The pattern is: `<runner> ' + localPrefix + '/hooks/<hook>'`
+      // or the old broken pattern: `<runner> ' + dirName + '/hooks/<hook>'`
+      const hookEscaped = hook.replace(/\./g, '\\.');
+      const brokenPattern = new RegExp(
+        `['"](?:node|bash)\\s['"]\\s*\\+\\s*dirName\\s*\\+\\s*['"]/hooks/${hookEscaped}['"]`
+      );
+      assert.ok(
+        !brokenPattern.test(src),
+        `${hook} must not use bare dirName — should use localPrefix for cwd-independent resolution`
+      );
+    });
+  }
+
+  test('no local hook command uses bare dirName + /hooks/', () => {
+    // Broader check: no local (non-global) hook path should use dirName directly
+    // The pattern `': '<runner> ' + dirName + '/hooks/'` is the broken form
+    const lines = src.split('\n');
+    const offenders = [];
+    for (let i = 0; i < lines.length; i++) {
+      const line = lines[i];
+      // Match lines that build local hook commands with bare dirName
+      if (/['"](?:node|bash)\s['"][^;]*\+\s*dirName\s*\+\s*['"]\/hooks\//.test(line)) {
+        offenders.push(`line ${i + 1}: ${line.trim()}`);
+      }
+    }
+    assert.equal(offenders.length, 0,
+      'Found local hook commands using bare dirName instead of localPrefix:\n' +
+      offenders.join('\n'));
+  });
+});

